storage_v2/spanstore/reader.go

```diff
@@ -0,0 +1,65 @@
+// Copyright (c) 2024 The Jaeger Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+package spanstore
+
+import (
+	"context"
+	"time"
+
+	"github.com/jaegertracing/jaeger/model"
+)
+
+// Reader finds and loads traces and other data from storage.
+type Reader interface {
+	// GetTrace retrieves the trace with a given id.
+	//
+	// If no spans are stored for this trace, it returns ErrTraceNotFound.
+	GetTrace(ctx context.Context, traceID model.TraceID) (*model.Trace, error)
+
+	// GetServices returns all service names known to the backend from spans
+	// within its retention period.
+	GetServices(ctx context.Context) ([]string, error)
+
+	// GetOperations returns all operation names for a given service
+	// known to the backend from spans within its retention period.
+	GetOperations(ctx context.Context, query OperationQueryParameters) ([]Operation, error)
+
+	// FindTraces returns all traces matching query parameters. There's currently
+	// an implementation-dependent abiguity whether all query filters (such as
+	// multiple tags) must apply to the same span within a trace, or can be satisfied
+	// by different spans.
+	//
+	// If no matching traces are found, the function returns (nil, nil).
+	FindTraces(ctx context.Context, query *TraceQueryParameters) ([]*model.Trace, error)
```

Comment:
hm, this one is tricky. We want the result to be `ptrace` model, but in `ptrace` there is not clear separation of spans from different traces, instead the separation is by resource/scope, which are not very relevant on retrieval (usually by then they are de-normalized onto individual spans in storage, although some storage backends can theoretically store fully normalized model). Plus, in order to reproduce the grouping of spans into resource/scope hierarchy we need to do relatively expensive comparison (fwiw the query service does this today for Process deduping, even though the UI does not make use of such normalization).

I believe in my RFC I proposed returning `[]*ptrace.TraceData` from these. The deduping could be an optional step/enrichment in the query-service, rather than forcing each storage backend do that.